sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc

```diff
@@ -0,0 +1,510 @@
+  });
+  ...
+}
+----
+
+The section below and the table that follows, describe the constructors, member functions and factory methods for `annotated_ptr`.
+The section below refers to an `annotated_ref` class which is described in the section following this one.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+template <typename T, typename PropertyListT = property_list_t<>>
+class annotated_ptr {
+  public:
+    using reference = annotated_ref<T, PropertyListT>;
+
+    annotated_ptr() noexcept = default;
+    explicit annotated_ptr(T *Ptr) noexcept = default;
+    annotated_ptr(annotated_ptr const &) noexcept = default;
+
+    // Conversion function
+    template <class U, class P> explicit annotated_ptr(annotated_ptr<U, P> const &) noexcept;
+
+    reference operator*() const noexcept;
+    reference operator[](std::ptrdiff_t) const noexcept;
+    annotated_ptr operator+(unsigned long long) const noexcept;
+    std::ptrdiff_t operator-(annotated_ptr) const noexcept;
+
+    operator bool() const noexcept;
+
+    // Implicit conversion is not supported
+    operator T*() noexcept = delete;
+    operator const T*() const noexcept = delete;
+
+    T* get() noexcept;
+    const T* get() const noexcept;
+
+    annotated_ptr& operator=(const T*) noexcept;
+    annotated_ptr& operator=(annotated_ptr const&) noexcept;
+
+    annotated_ptr& operator++() noexcept;
+    annotated_ptr operator++(int) noexcept;
+    annotated_ptr& operator--() noexcept;
+    annotated_ptr operator--(int) noexcept;
+
+    template<typename propertyT>
+    static constexpr bool has_property();
+
+    // The return type is an unspecified internal class used to represent
+    // instances of propertyT
+    template<typename propertyT>
+    static constexpr /*unspecified*/ get_property();
+
+  private:
+    T *Ptr;
+  };
+} // namespace sycl::ext::oneapi
+
+----
+
+[frame="topbot",options="header"]
+|===
+|Functions |Description
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr() noexcept;
+----
+|
+Constructs an `annotated_ptr` object. Does not allocate new storage. The underlying pointer is initialized to `nullptr`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+explicit annotated_ptr(T *Ptr);
+----
+|
+Constructs an `annotated_ptr` object. Does not allocate new storage. The underlying pointer is initialized with `Ptr`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr(annotated_ptr const &) noexcept = default;
+----
+|
+Constructs an `annotated_ptr` object from another `annotated_ptr` with the same template parameterization object.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template <class U, class P> explicit annotated_ptr(annotated_ptr<U, P> const & ConvertFrom);
+----
+|
+Constructs the `annotated_ptr` object from the `ConvertFrom` object if the template parameter types are compatible. For example, if the properties on one of the operands conflict with the others then the compiler should issue a compile time error.
```

Comment:
This wording seems to imply that it is OK to remove a property from `ConvertFrom` when constructing the `this` object.  That seems wrong to me.  It seems like it should be an error unless the `PropertyListT` template parameter is a superset of the `P` template parameter.

In addition, do we really want to support the case when `T` and `U` are different?

Reply:
Corrected the wording and no I do not think we need to support the case where T and U are different. Updated.